lib/recursive-methods.rb

```diff
@@ -1,49 +1,92 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_recursive(s, reversed_string = "", index = -1)
+end
+
+def reverse_recursive(s, reversed_string, index)
+  return reversed_string if index < -s.length
+  reversed_string += s[index]
+  return reverse_recursive(s, reversed_string, index - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return reverse_inplace_recursive(s, last = s.length - 1, index = 0)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_inplace_recursive(s, last, index)
+  return s if index >= last
+  current_letter = s[index]
+  s[index] = s[last]
+  s[last] = current_letter
+  reverse_inplace_recursive(s, last - 1, index + 1)
+end
+
+# Time complexity: o(n)
+# Space complexity: o(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n <= 0
+  return 2 if n == 1
+  return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.empty?
+  return false if s[0] != "(" || s[-1] != ")"
+  return nested(s[1...-1])
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.